### PR TITLE
added schema validation for update

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

File: modules/crud/schema/schemaUpdate.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		return nil
+	}
+
+	schemaDb, ok := s.SchemaDoc[dbType]
+	if !ok {
+		return errors.New(dbType + " Db Not present in Schema")
+	}
+	schemaDoc, ok := schemaDb[col]
+	if !ok {
+		return errors.New(col + " Col Not present in Schema of " + dbType)
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: Return nil here

File: modules/crud/schema/schemaUpdate.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+			for index, value := range t {
+				newVal, err := s.checkType(value, schemaDocValue)
+				if err != nil {
+					return nil, err
+				}
+				arr[index] = newVal
+			}
+			return arr, nil
+		default:
+			return nil, errors.New("Schema Update Math Op. Wrong type ")
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: Update array not math

File: modules/crud/schema/schemaUpdate.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+	v, ok := doc.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("Schema math op wrong type passed expecting map[string]interface{}")
+	}
+	newMap := map[string]interface{}{}
+	var temp interface{}
+	for fieldKey, fieldValue := range v {
+
+		schemaDocValue, ok := schemaDoc[fieldKey]
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: All you need to do is check if the kind is date time, thats it. No need to do anything else

File: modules/crud/schema/schemaUpdate.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+	for fieldKey, fieldValue := range v {
+
+		schemaDocValue, ok := schemaDoc[fieldKey]
+		if !ok {
+			return nil, errors.New("field not found in schemaField")
+		}
+
+		switch t := fieldValue.(type) {
+		case []interface{}:
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: Need not be an array all the time

File: modules/crud/schema/schemaUpdate.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		case "currentDate", "currentTimestamp":
+			temp, err = validateDateOperations(doc, schemaDoc)
+			if err != nil {
+				return err
+			}
+		default:
+			return errors.New(key + " Update operator not supported")
+		}
+		updateDoc[key] = temp
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: You need not update this at all. You just need to validate

File: modules/crud/schema/schemaUpdate.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	for fieldKey, fieldValue := range v {
+
+		schemaDocValue, ok := schemaDoc[fieldKey]
+		if !ok {
+			return errors.New("field not found in schemaField")
+		}
+
+		switch fieldValue.(type) {
+		case int:
+			switch schemaDocValue.Kind {
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: Instead of switch case use type assert

File: modules/crud/schema/schemaUpdate.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+			case TypeInteger:
+			default:
+				return errors.New("Integer wrong type wanted Integer")
+			}
+		case float32, float64:
+			switch schemaDocValue.Kind {
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: Instead of switch case use type assert

File: modules/crud/schema/schemaUpdate.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+			case TypeFloat:
+			default:
+				return errors.New("Float wrong type wanted Float")
+			}
+		default:
+			return errors.New("Schema Update Math Op. Wrong type ")
```
Comment (YourTechBud) [inline modules/crud/schema/schemaUpdate.go]: All error messages should not have capital letters

File: modules/crud/schema/schemaUpdate.go (hunk 5 above)
Comment (YourTechBud) [reply modules/crud/schema/schemaUpdate.go]: Only update for the set clause

File: modules/crud/schema/schemaUpdate.go (hunk 7 above)
Comment (sharadregoti) [reply modules/crud/schema/schemaUpdate.go]: first of all why type assert?
& for float which type float32 or 64

Comment (YourTechBud) [reply modules/crud/schema/schemaUpdate.go]: My bad. I meant if else

**Comment by sharadregoti** [comment]
schema inspection:
gives the output as a schema doc structure,
converting it to SDL is still remaining & some minor bug fixes

File: modules/auth/auth.go [hunk 9]
```diff
@@ -50,6 +53,16 @@ func (m *Module) SetConfig(project string, secret string, rules config.Crud, fil
 	if pubsub != nil && pubsub.Enabled {
 		m.pubsubRules = pubsub.Rules
 	}
+
+	s := schema.Init(m.crud)
```
Comment (YourTechBud) [inline modules/auth/auth.go]: No need to Init again here

File: modules/auth/auth.go [hunk 10]
```diff
@@ -50,6 +53,16 @@ func (m *Module) SetConfig(project string, secret string, rules config.Crud, fil
...
+	if err := s.ParseSchema(m.rules); err != nil {
+		return err
+	}
+	m.schema = s
```
Comment (YourTechBud) [inline modules/auth/auth.go]: this stem isn't required

File: modules/auth/schema/inspection.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	inspectionDb[dbType] = inspectionCollection
+
+	return inspectionDb, nil
+}
+
+func inspectionCheckFieldType(typeName string, fieldDetails *SchemaFieldType) error {
+	// TODO: what about my-sql set type
+	// TODO: sql types int(11), varchar(255) is there any thing else
+	switch typeName {
+	case "char", "varchar(255)", "tinytext", "text", "blob", "mediumtext", "mediumblob", "longtext", "longblob", "decimal":
```
Comment (YourTechBud) [inline modules/auth/schema/inspection.go]: what if its varchar(50) or something. You should ideally just check the prefixes for int and varchar

File: modules/auth/schema/inspection.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+			fieldDetails.IsList = true
+		}
+		// check foreignKey & identify if relation exists
+		for _, foreignValue := range foreignkeys {
+			if foreignValue.ColumnName == value.FieldName {
+				fieldDetails.TableJoin = foreignValue.RefTableName
+				fieldDetails.Kind = TypeJoin
+				fieldDetails.Directive.Kind = "relation"
+				// TODO: recuresive function for nested object
+				result, err := s.schemaInspection(ctx, dbType, project, foreignValue.RefTableName)
```
Comment (YourTechBud) [inline modules/auth/schema/inspection.go]: Nested object aren't possible in MySQL. Remove this and the todo comment

File: modules/auth/schema/update.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	if !ok {
+		return nil, errors.New("Schema update set op wrong type passed expecting map[string]interface{}")
+	}
+
+	newMap := map[string]interface{}{}
+	for key, value := range v {
+		// check if key present in schemaDoc if not insert the field
+		schemaDocValue, ok := schemaDoc[key]
+		if !ok {
+			return nil, errors.New("Scheam set op. field not found in schemaField")
```
Comment (YourTechBud) [inline modules/auth/schema/update.go]: the error message seems to have typos. Fix that

File: modules/auth/schema/update.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+	for fieldKey, fieldValue := range v {
+
+		schemaDocValue, ok := schemaDoc[fieldKey]
+		if !ok {
+			return errors.New("field not found in schemaField")
+		}
+
+		switch fieldValue.(type) {
+		case int:
+			if schemaDocValue.Kind != TypeInteger {
```
Comment (YourTechBud) [inline modules/auth/schema/update.go]:
```suggestion
			if schemaDocValue.Kind != TypeInteger && schemaDocValue.Kind != TypeFloat {
```

File: modules/auth/schema/inspection.go (hunk 11 above)
Comment (YourTechBud) [reply modules/auth/schema/inspection.go]: Answered offline

File: modules/auth/schema/update.go (hunk 14 above)
Comment (YourTechBud) [reply modules/auth/schema/update.go]: Answered offline

**Pull request merged**